src/toucan2/jdbc/read.clj

```diff
@@ -128,18 +128,12 @@
   [_conn _model rset _rsmeta i]
   (get-object-of-class-thunk rset i java.time.OffsetTime))
 
-(defn- make-column-thunk [conn model ^ResultSet rset i]
-  (log/tracef "Building thunk to read column %s" i)
-  (fn column-thunk []
-    (let [rsmeta (.getMetaData rset)
-          thunk  (read-column-thunk conn model rset rsmeta i)
-          v      (thunk)]
-      (next.jdbc.rs/read-column-by-index v rsmeta i))))
-
-(defn- make-i->thunk [conn model rset]
-  (comp (memoize (fn [i]
-                   (make-column-thunk conn model rset i)))
-        int))
+(defn- read-column-value [conn model ^ResultSet rset i]
+  (let [i (int i)
+        rsmeta (.getMetaData rset)
+        thunk  (read-column-thunk conn model rset rsmeta i)
```

Comment:
So it is actually somewhat important here to calculate these thunks just once for the entire result set instead of once per value. The overhead of calling the `read-column-thunk` multimethod on every single value in the results can get pretty high.

This whole `read-column-thunk` idea came about when I did something similar in Metabase.  Some of these `read-column-thunk` methods have to look at `ResultSetMetaData` to determine how to read a column out or make other decisions like that that only need to happen once per ResultSet rather than once per row. Here's an example with Postgres in Metabase:

https://github.com/metabase/metabase/blob/f49407fcbdf18f40e247437f41b3bbefd1e227bf/src/metabase/driver/postgres.clj#L795-L803

Both `timestamp` and `timestamp with time zone` come back as `java.sql.Types/TIMESTAMP` so you need to look at the `ResultSetMetaData` to determine what the actual column type is.

Reply:
I understand your intent and agree that it would be beneficial. The problem is, if I'm reading the code correctly, the process that you describe is not actually cached right now. Obtaining metadata and calling `read-column-thunk` happens here https://github.com/camsaul/toucan2/blob/master/src/toucan2/jdbc/read.clj#L134, inside the cached lambda, and so it happens for each row.

I can try to rewrite this code so that it is properly cached and reused.